File: leetcode/tree/avl/avl_tree.cpp
// AVL Tree Implementation
#include <algorithm>
#include <cassert>
using namespace std;

struct Node {
    int key, height; Node *left, *right;
    Node(int k) : key(k), height(1), left(nullptr), right(nullptr) {}
};

class AVLTree {
    int h(Node* n) { return n ? n->height : 0; }
    void upd(Node* n) { n->height = 1 + max(h(n->left), h(n->right)); }
    Node* rotR(Node* y) { Node* x = y->left; y->left = x->right; x->right = y; upd(y); upd(x); return x; }
    Node* rotL(Node* x) { Node* y = x->right; x->right = y->left; y->left = x; upd(x); upd(y); return y; }
public:
    Node* insert(Node* n, int key) {
        if (!n) return new Node(key);
        if (key < n->key) n->left = insert(n->left, key);
        else if (key > n->key) n->right = insert(n->right, key);
        else return n;
        upd(n); int b = h(n->left) - h(n->right);
        if (b > 1 && key < n->left->key) return rotR(n);
        if (b < -1 && key > n->right->key) return rotL(n);
        if (b > 1 && key > n->left->key) { n->left = rotL(n->left); return rotR(n); }
        if (b < -1 && key < n->right->key) { n->right = rotR(n->right); return rotL(n); }
        return n;
    }
};

int main() {
    AVLTree avl; Node* root = nullptr;
    for (int k : {10,20,30,40,50,25}) root = avl.insert(root, k);
    return 0;
}

File: leetcode/tree/avl/avl_tree.py
# AVL Tree Implementation
class Node:
    def __init__(self, key):
        self.key = key; self.left = self.right = None; self.height = 1

class AVLTree:
    def h(self, n): return n.height if n else 0
    def bal(self, n): return self.h(n.left) - self.h(n.right) if n else 0
    def rotR(self, y):
        x = y.left; y.left = x.right; x.right = y
        y.height = 1 + max(self.h(y.left), self.h(y.right))
        x.height = 1 + max(self.h(x.left), self.h(x.right)); return x
    def rotL(self, x):
        y = x.right; x.right = y.left; y.left = x
        x.height = 1 + max(self.h(x.left), self.h(x.right))
        y.height = 1 + max(self.h(y.left), self.h(y.right)); return y
    def insert(self, node, key):
        if not node: return Node(key)
        if key < node.key: node.left = self.insert(node.left, key)
        elif key > node.key: node.right = self.insert(node.right, key)
        else: return node
        node.height = 1 + max(self.h(node.left), self.h(node.right))
        b = self.bal(node)
        if b > 1 and key < node.left.key: return self.rotR(node)
        if b < -1 and key > node.right.key: return self.rotL(node)
        if b > 1 and key > node.left.key: node.left = self.rotL(node.left); return self.rotR(node)
        if b < -1 and key < node.right.key: node.right = self.rotR(node.right); return self.rotL(node)
        return node

if __name__ == "__main__":
    avl = AVLTree(); root = None
    for k in [10,20,30,40,50,25]: root = avl.insert(root, k)
    print("AVL OK")
